subprojects/distributions/src/integTest/groovy/org/gradle/BinaryBreakageIntegrationTest.groovy

```diff
@@ -0,0 +1,133 @@
+
+        'org.gradle.api.invocation.Gradle': '''***! MODIFIED INTERFACE: PUBLIC ABSTRACT org.gradle.api.invocation.Gradle  (not serializable)
+\t***! MODIFIED METHOD: PUBLIC ABSTRACT org.gradle.includedbuild.IncludedBuild (<-org.gradle.api.initialization.IncludedBuild) includedBuild(java.lang.String)''',
+        'org.gradle.StartParameter': '''***! MODIFIED CLASS: PUBLIC org.gradle.StartParameter  (default serialVersionUID changed)
+\t---! REMOVED INTERFACE: org.gradle.initialization.ParallelismConfiguration'''
+    ]
+
+    private List<PathMatcher> publicApiIncludeMatchers
+
+    private List<PathMatcher> publicApiExcludeMatchers
+
+    def setup() {
+        publicApiIncludeMatchers = PUBLIC_API_INCLUDES.collect { PatternMatcherFactory.compile(true, it) }
+        publicApiExcludeMatchers = PUBLIC_API_EXCLUDES.collect { PatternMatcherFactory.compile(true, it) }
+        assert !publicApiIncludeMatchers.empty
+        assert !publicApiExcludeMatchers.empty
+    }
+
+    def "should be binary compatibility with Gradle #version"() {
+        GradleDistribution current = new UnderDevelopmentGradleDistribution()
+
+        // For a hotfix release we need to trust our own judgement or run this test manually
+        assumeTrue(previous.version < current.version)
+
+        when:
+        def comparison = comparePublicApi(previous, current)
+
+        then:
+        def binaryIncompatibleChanges = comparison.findAll { !it.binaryCompatible }.collectEntries { JApiClass apiClass ->
+            [(apiClass.fullyQualifiedName): describeBinaryBreakingChange(apiClass)]
+        }
+        def remainingChanges = binaryIncompatibleChanges - ACCEPTED_CHANGES
+        remainingChanges.values().isEmpty()
+
+
+        where:
+        previous << [new ReleasedVersionDistributions(buildContext).mostRecentFinalRelease]
+        version = previous.version.version
+    }
+
+    private static String describeBinaryBreakingChange(JApiClass binaryIncompatibleChange) {
+        def outputOptions = Options.newDefault()
+        outputOptions.outputOnlyBinaryIncompatibleModifications = true
+        new StdoutOutputGenerator(outputOptions, [binaryIncompatibleChange]).generate().readLines().tail().join('\n')
+    }
+
+    private List<JApiClass> comparePublicApi(GradleDistribution previous, GradleDistribution current) {
+        def options = new JarArchiveComparatorOptions()
+
+        def currentJars = getDistributionJars(current)
+        def oldJars = getDistributionJars(previous)
+
+        options.classPathMode = JarArchiveComparatorOptions.ClassPathMode.TWO_SEPARATE_CLASSPATHS
+        options.newClassPath = currentJars.external*.absolutePath as List
+        options.oldClassPath = oldJars.external*.absolutePath as List
+
+        def oldGradleJars = oldJars.internal.collect { new JApiCmpArchive(it, previous.version.version) }
+        def currentGradleJars = currentJars.internal.collect { new JApiCmpArchive(it, buildContext.version.version) }
```

Comment:
Yes, this can be a separate step. It just means that Groovy/Kotlin/Junit updates are inherently more risky, because we don't yet auto-check their compatibility.